### PR TITLE
Iterating on Simulator Design for Sampling

### PR DESCRIPTION
**Pull request by sreekaroo**
# Overview

Focused on high-level design for the simulator. Modularized parts of the simulator so we can iterate on designs more easily. This draft PR is mostly just to share code easily. The goal was to create a simple data ingestor and sampler that could sample by letter from multiple data sources. View `sim_test.py` for examples of use.

I think the replay session logic will fit into this design without any change as well. The sampler would just be an iterator, that I implemented in the previous design. The DataEngine would load in-session data instead, which is also already implemented.

## Ticket

Link a pivotal ticket here

## Contributions

- Created interfaces for `DataEngine`, `Sampler`, `MetricReferee`, `ModelHandler`, `StateManager`
- Implemented `RawDataEngine` and `SimpleLetterSampler`, which can successfully load in multiple data sources, organize into queryable schema, and return eeg samples that can be fed into SignalModel
- Implemented draft `StateManager` that handles a lot of the updating logic from inquiry to inquiry

### Started but unfinished
- The simulator run loop needs a little more work, view in `sim_sampler.py`
- `MetricReferee`, `ModelHandler`,  all have unfinished implementations
- `StateManager` is a bit naive right now. There is also an issue with how I'm evaluating evidence because the simulator is always making a decision; albeit, mostly wrong ones.

## Test

-No unit testing yet, just ran script to test sampling and data loading in `sim_test.py`

## Documentation

- Are documentation updates required? In-line, README, or [documentation](https://github.com/BciPy/bcipy.github.io)? Verify the updates you did here.

## Changelog

- Is the CHANGELOG.md updated with your detailed changes?

### CODE REVIEW

File: bcipy/simulator/helpers/data_engine.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    def load(self) -> DataEngine:
+        """
+        Processes raw data from data folders using provided parameter files.
+            - Extracts and stores trial data, stimuli, and stimuli_labels by inquiries
+
+        Returns:
+            self for chaining
+        """
+        assert len(self.source_dirs) == len(self.parameters)
```
Comment (lawhead) [inline bcipy/simulator/helpers/data_engine.py]: We may want a more general validation method here. Consider when, or if, it's okay for parameters to be different. For example, it may make sense for data to have different filters if it's collected in different places (AC electrical frequency in US is 60hz, but may be different elsewhere).

File: bcipy/simulator/helpers/data_engine.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+
+class RawDataEngine(DataEngine):
+    """
+    Object that loads in list of previous session data folders and transforms data into implemented schema
+    """
+
+    def __init__(self, source_dirs: list[str], parameters: list[Parameters]):
+        self.source_dirs: list[str] = source_dirs
+        self.parameters: list[Parameters] = parameters
```
Comment (lawhead) [inline bcipy/simulator/helpers/data_engine.py]: A simpler API may be to have each source_dir contain its own parameters file.

Also, `parameters` is used throughout the BciPy code as the name for the contents of a single parameters.json file. It's technically correct but a little confusing to read this as a list of those files. I don't have any great ideas for alternatives, though.

File: bcipy/simulator/helpers/data_engine.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        The "eeg" data for trial looks like this
+            - ndarray.shape = (channel_n, sample_n)
+            - a trial is an eeg response for a symbol being displayed
+
+        Returns:
+            self for chaining
+        """
+
+        cols = ["series_n", "inquiry_n", "trial_n", "symbol", "target", "eeg"]
+        types = [int, int, int, str, int, np.ndarray]
```
Comment (lawhead) [inline bcipy/simulator/helpers/data_engine.py]: Another alternative would be to use a NamedTuple as the schema. Then it could be returned even before calling `transform`.

File: bcipy/simulator/helpers/data_engine.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+    def __init__(self, source_dirs: list[str], parameters: list[Parameters]):
+        self.source_dirs: list[str] = source_dirs
+        self.parameters: list[Parameters] = parameters
+        self.data: Optional[list[ExtractedExperimentData]] = None
+        self.trials_by_inquiry: list[np.ndarray] = []  # shape (i_inquiry, n_channel, m_trial, x_sample).
+        self.symbols_by_inquiry: list[list] = []  # shape (i_inquiry, s_alphabet_subset)
+        self.labels_by_inquiry: list[list] = []  # shape (i_inquiry, s_alphabet_subset)
+        self.schema: Optional[pd.DataFrame] = None
+        self.load()
```
Comment (lawhead) [inline bcipy/simulator/helpers/data_engine.py:67]: I would use `List` (from typing import List) for all of the type annotations.

File: bcipy/simulator/helpers/sampler.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    def set_reshaper(self, reshaper: Callable):
+        ...
+
+
+class SimpleLetterSampler(Sampler):
+
+    def __init__(self, data_engine: RawDataEngine, params=None):
+        self.data_engine: RawDataEngine = data_engine
+        self.params: Parameters = params if params else self.data_engine.parameters
```
Comment (lawhead) [inline bcipy/simulator/helpers/sampler.py]: I think the type annotation is not correct here. `self.data_engine.parameters` is `List[Parameters]`.

File: bcipy/simulator/helpers/sampler.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+        """
+
+        inquiry_letter_subset = state.display_alphabet
+        target_letter = state.target_symbol
+        eeg_responses = []
+        for symbol in inquiry_letter_subset:
+            is_target = int(symbol == target_letter)
+            filtered_data = self.data.query(f'target == {is_target} and symbol == "{symbol}"')
+            row = filtered_data.sample(
+                1) if len(filtered_data) else None  # TODO figure out what to do when no sample is found. e.g no eeg data for nontarget symbol of 'x'
```
Comment (lawhead) [inline bcipy/simulator/helpers/sampler.py]: My first instinct is to raise an exception here, since this class doesn't have the ability to generate new data. That suggests that we may want a validation function in the constructor to ensure that the underlying data is sufficient so we can fail as early as possible.

File: bcipy/simulator/helpers/sim_state.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+@dataclass
+class SimSessionState:
+    """ Represents the state of a current session during simulation """
+    target_symbol: str
+    current_sentence: str
+    target_sentence: str
+    display_alphabet: list[str]
+    inquiry_n: int
+    series_n: int
```
Comment (lawhead) [inline bcipy/simulator/helpers/sim_state.py]: I like the idea of capturing this all in one state object that can be passed around. We will probably want to add the lists of probabilities. Check out the `DecisionMaker` `list_series` property in task/control/handler.py.

File: bcipy/simulator/tests/sampler_test.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    display_alp = random.sample(alphabet(), 10)
+    state: SimSessionState = SimSessionState(target_symbol=target_symbol, inquiry_n=0, series_n=0, target_sentence=target_phrase,
+                                             current_sentence="", display_alphabet=display_alp)
+
+    sampler: Sampler = SimpleLetterSampler(data_engine)
+    sample: np.ndarray = sampler.sample(state)
+
+    model = PcaRdaKdeModel()
+    model = model.load(model_file)
```
Comment (lawhead) [inline bcipy/simulator/tests/sampler_test.py]: Just as an FYI, model loading has changed with the multimodal work and newer data may be different that what you have. Now the entire model class is pickled, not just its internal contents. So this method of first creating a `PcaRdaKdeModel` and then having it load the pickle file has changed to `pickle.load(model_file)`. We're still evaluating if this is a good approach, though.

**Review by lawhead — COMMENTED** [review]
This is great progress and seems to be heading in the right direction.

File: bcipy/simulator/helpers/sampler.py (hunk 5 above)
Comment (sreekaroo) [reply bcipy/simulator/helpers/sampler.py]: yeah good catch

File: bcipy/simulator/helpers/data_engine.py (hunk 3 above)
Comment (sreekaroo) [reply bcipy/simulator/helpers/data_engine.py]: not sure this will work with the querying logic that dataframes have to offer

File: bcipy/simulator/helpers/data_engine.py (hunk 2 above)
Comment (sreekaroo) [reply bcipy/simulator/helpers/data_engine.py]: Yes will refactor for this. I'll brainstorm more explicit naming

File: bcipy/simulator/helpers/data_engine.py (hunk 3 above)
Comment (lawhead) [reply bcipy/simulator/helpers/data_engine.py]: Mostly I was talking about the rows. The rows could be represented as a named tuple rather than a dict. The main advantage is that the fields and their types are self-documented. Pandas can create a DataFrame from a list of tuples.

Also, `cols` and `types` are unused.

Comment (sreekaroo) [reply bcipy/simulator/helpers/data_engine.py]: oh i see your point.  yeah i left that there accidentally

File: bcipy/simulator/helpers/decision.py [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+    def decide(self, series: List[InquiryResult]):
+        current_distribution = series[-1].evidence_likelihoods
+        if np.max(current_distribution) > self.tau:
+            log.info("Committing to decision: posterior exceeded threshold.")
+            return True
+        return False
+
+#
+# class MarginCriteria(DecisionCriteria):
```
Comment (lawhead) [inline bcipy/simulator/helpers/decision.py]: I would definitely not include Margin or Momentum criteria at this point. They have not yet been used in BciPy

File: bcipy/simulator/helpers/sim_state.py [hunk 10]
```diff
@@ -1,28 +1,129 @@
...
+
+    def __init__(self, parameters: Parameters):
+        self.state: SimState = self.initial_state()
+        self.parameters = parameters
+
+        self.stop_inq = 50  # TODO pull from parameters
+
+    def is_done(self) -> bool:
+
+        return self.state.total_inquiry_count() > self.stop_inq or self.state.target_sentence == self.state.current_sentence or self.state.series_n > 50
```
Comment (lawhead) [inline bcipy/simulator/helpers/sim_state.py]: It may be useful to log what caused the simulation to conclude.

File: bcipy/simulator/helpers/sim_state.py [hunk 11]
```diff
@@ -1,28 +1,129 @@
...
+
+        temp_inquiry_result = InquiryResult(target=self.state.target_symbol, time_spent=0, stimuli=self.state.display_alphabet,
+                                            evidence_likelihoods=list(evidence), decision=None)
+
+        # finding out whether max iterations is hit or prob threshold is hit
+        temp_series = copy.deepcopy(self.get_state().series_results)
+        temp_series[-1].append(temp_inquiry_result)
+        is_decidable = any([decider.decide(temp_series[-1]) for decider in self.state.decision_criterion])
+        decision = None
+        # TODO what to do when max inquiry count is reached?
```
Comment (lawhead) [inline bcipy/simulator/helpers/sim_state.py]: Maximum number of inquiries should probably be part of the definition of `is_done`. You may want an assertion at the top of this method to ensure that `not self.is_done()`.

File: bcipy/simulator/helpers/sim_state.py [hunk 12]
```diff
@@ -1,28 +1,129 @@
...
+        self.stop_inq = 50  # TODO pull from parameters
+
+    def is_done(self) -> bool:
+
+        return self.state.total_inquiry_count() > self.stop_inq or self.state.target_sentence == self.state.current_sentence or self.state.series_n > 50
+
+    def update(self, evidence: np.ndarray) -> InquiryResult:
+
+        temp_inquiry_result = InquiryResult(target=self.state.target_symbol, time_spent=0, stimuli=self.state.display_alphabet,
+                                            evidence_likelihoods=list(evidence), decision=None)
```
Comment (lawhead) [inline bcipy/simulator/helpers/sim_state.py]: `time_spent` is going to be a computed value based on the parameters for the number of stim, how much time per stim, fixation time, etc. See the `bcipy/helpers/acquisition.py` module for the calculation that we use to compute the data buffer size in seconds. Maybe factor this out into a method to calculate this value.

Also, we should make a note somewhere for refactoring needed when we discuss multimodal. This method will likely need to accept evidences for each device and probably the same for InquiryResult.

File: bcipy/simulator/helpers/sim_state.py [hunk 13]
```diff
@@ -1,28 +1,129 @@
...
+
+        # finding out whether max iterations is hit or prob threshold is hit
+        temp_series = copy.deepcopy(self.get_state().series_results)
+        temp_series[-1].append(temp_inquiry_result)
+        is_decidable = any([decider.decide(temp_series[-1]) for decider in self.state.decision_criterion])
```
Comment (lawhead) [inline bcipy/simulator/helpers/sim_state.py]: We should log why a decision was made.

File: bcipy/simulator/helpers/sim_state.py [hunk 14]
```diff
@@ -1,28 +1,129 @@
...
+                next_target_symbol_idx = len(new_state['current_sentence'])
+                if next_target_symbol_idx >= len(self.state.target_sentence):
+                    pass
+                else:
+                    new_state['target_symbol'] = self.state.target_sentence[next_target_symbol_idx]
+
+            else:  # wrong decision
+                new_state['series_n'] += 1
+                new_state['series_results'].append([])
+                new_state['inquiry_n'] = 0
```
Comment (lawhead) [inline bcipy/simulator/helpers/sim_state.py]: In the case that the spelled letter does not match the target, it it still displayed as part of the current sentence. However, then the next target becomes the backspace character. See the `RSVPCopyPhraseTask` class `next_target` method. You probably don't need an if-else here. I think the only thing that would be different is the selection of the next target.

File: bcipy/simulator/helpers/types.py [hunk 15]
```diff
@@ -0,0 +1,11 @@
+from dataclasses import dataclass
+from typing import Optional, List
+
+
+@dataclass
+class InquiryResult:
+    target: Optional[str]
+    time_spent: int  # TODO what does time_spent mean?
```
Comment (lawhead) [inline bcipy/simulator/helpers/types.py]: Total number of seconds spent in an inquiry. This should be a float.

File: bcipy/simulator/helpers/types.py [hunk 16]
```diff
@@ -0,0 +1,11 @@
...
+    stimuli: List
+    evidence_likelihoods: List
```
Comment (lawhead) [inline bcipy/simulator/helpers/types.py]: Not all evidence produces likelihoods. Language model and button press are probabilities.

File: bcipy/simulator/helpers/sim_state.py [hunk 17]
```diff
@@ -1,28 +1,129 @@
...
+            state_dict[state_field] = state_value
+            self.state = SimState(**state_dict)
+            return self.get_state()
+        else:
+            raise FieldException(f"Cannot find state field {state_field}")
+
+    @staticmethod
+    def initial_state(parameters: Parameters = None) -> SimState:
+        sentence = "HELLO_WORLD"  # TODO abstract out with sim_parameters.json
+        target_symbol = sentence[0]
```
Comment (lawhead) [inline bcipy/simulator/helpers/sim_state.py]: We also have a parameter for the number of letters typed so the user can start in the middle of a word or phrase. This can be added later, though.

File: bcipy/simulator/helpers/types.py (hunk 16 above)
Comment (sreekaroo) [reply bcipy/simulator/helpers/types.py]: ok maybe as we incorporate other models, I'll turn this into separate data structure
